### PR TITLE
Feat/fcgi socket monitoring

### PR DESCRIPTION
**Pull request by benoitdm-oslandia**
This follow the PR https://github.com/qgis/QGIS/pull/51744

> This feature improves feedback/cancel management during the map rendering to handle when the client closes the in/out socket. This can occur when the client (an http server as nginx) has a timeout shorter than the rendering time (fe. the rendering is very long due database issue). Without this feature, the http server may resent new requests to the qgis server which is busy and can not handle it.
> 
> For each new FCGI request, we start a new thread to monitor the incoming socket stream and detect its closing. If the socket closes before the rendering, we update a feedback object to cancel the rendering jobs. When the rendering is done, the thread is closed.

Sponsored by Orange @eropartz

[X] should be backported to LTR

### CODE REVIEW

File: src/server/qgsfcgiserverresponse.cpp [hunk 1]
```diff
@@ -23,15 +23,137 @@
...
+#include "qgslogger.h"
+
+#ifdef Q_OS_LINUX
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#endif
+
+#ifdef Q_OS_LINUX
```
Comment (lbartoletti) [inline src/server/qgsfcgiserverresponse.cpp:36]:
```suggestion
#ifdef Q_OS_UNIX
```

File: src/server/qgsfcgiserverresponse.h [hunk 2]
```diff
@@ -26,6 +26,33 @@
...
+{
+    Q_OBJECT
+
+  public:
+
+    /**
+       * \brief SocketMonitoringThread
+       * \param  isResponseFinished
+       * \param  feedback
+       */
```
Comment (lbartoletti) [inline src/server/qgsfcgiserverresponse.h:47]: seems ill indented

File: src/server/qgsfcgiserverresponse.cpp [hunk 3]
```diff
@@ -23,15 +23,137 @@
...
+{
+  if ( ! mIpcFd )
+  {
+    QgsMessageLog::logMessage( QStringLiteral( "Socket monitoring disabled: no socket fd!" ),
+                               QStringLiteral( "FCGIServer" ),
+                               Qgis::MessageLevel::Warning );
+    return;
+  }
+
+#ifdef Q_OS_LINUX
```
Comment (lbartoletti) [inline src/server/qgsfcgiserverresponse.cpp:104]:
```suggestion
#ifdef Q_OS_UNIX
```

File: src/server/qgsfcgiserverresponse.cpp [hunk 4]
```diff
@@ -23,15 +23,137 @@
...
+SocketMonitoringThread::SocketMonitoringThread( bool *isResponseFinished, QgsFeedback *feedback )
+  : mIsResponseFinished( isResponseFinished )
+  , mFeedback( feedback )
+  , mIpcFd( -1 )
+{
+  setObjectName( "FCGI socket monitor" );
+  Q_ASSERT( mIsResponseFinished );
+  Q_ASSERT( mFeedback );
 
+#ifdef Q_OS_LINUX
```
Comment (lbartoletti) [inline src/server/qgsfcgiserverresponse.cpp:70]:
```suggestion
#ifdef Q_OS_UNIX
```

File: src/server/qgsfcgiserverresponse.h (hunk 2 above)
Comment (benoitdm-oslandia) [reply src/server/qgsfcgiserverresponse.h:47]: fixed

File: src/server/qgsfcgiserverresponse.cpp [hunk 5]
```diff
@@ -23,15 +23,137 @@
 #include <fcgi_stdio.h>
 #include <QDebug>
 
+#include <unistd.h>
```
[the remark this answers is not in the document]
Comment (pblottiere) [reply src/server/qgsfcgiserverresponse.cpp:26]: BTW, I dont think that `unistd.h` header is necessary for compiling.

**Comment by pblottiere** [comment]
Hello @benoitdm-oslandia,


First, for the sake of history and to make this long story short:

1. initial implementation with `SO_LINGER` option and with unix sockets only -> not portable (GNU/Linux only)
2. new implementation based on `QLocalSocket` and without `SO_LINGER` for full portability -> not stable (segfault, ...)
3. new implementation based on unix sockets but **without** `SO_LINGER` -> should be portable on all UNIX-like operating systems


I tested the new implementation and everything looks good :tada: :
  - the monitoring thread properly detects when the socket is closed by the web server and the feedback is canceled
  - there isn't any strange behavior (segfault, invalid response, ...)


I'm going to take a look on the code itself, but in the meantime, we're in feature freeze, so you need to update the documentation for 3.36.

File: src/server/qgsfcgiserverresponse.h [hunk 6]
```diff
@@ -26,6 +26,33 @@
...
 #include <QBuffer>
+#include <QThread>
+
+/**
+ * \ingroup server
+ * \class SocketMonitoringThread
+ * \brief Thread used to monitor the fcgi socket
+ * \since QGIS 3.34
+ */
+class SocketMonitoringThread: public QThread
```
Comment (pblottiere) [inline src/server/qgsfcgiserverresponse.h:37]:
```suggestion
class QgsSocketMonitoringThread: public QThread
```

File: src/server/qgsfcgiserverresponse.h [hunk 7]
```diff
@@ -26,6 +26,33 @@
...
+       * \param  isResponseFinished
+       * \param  feedback
+       */
+    SocketMonitoringThread( bool *isResponseFinished, QgsFeedback *feedback );
+    void run( );
+
+  private:
+    bool *mIsResponseFinished = nullptr;
+    QgsFeedback *mFeedback = nullptr;
+    int mIpcFd = 0;
```
Comment (pblottiere) [inline src/server/qgsfcgiserverresponse.h:54]:
```suggestion
    int mIpcFd = -1;
```

File: src/server/services/wms/qgsmaprendererjobproxy.h [hunk 8]
```diff
@@ -53,8 +53,10 @@ namespace QgsWms
        * Sequential or parallel map rendering.
        * \param mapSettings Passed to MapRendererJob
        * \param image The resulting image
+       * \param feedback socket feedback
```
Comment (pblottiere) [inline src/server/services/wms/qgsmaprendererjobproxy.h:56]:
```suggestion
       * \param feedback Socket feedback
```

File: src/server/services/wms/qgsmaprendererjobproxy.h [hunk 9]
```diff
@@ -53,8 +53,10 @@ namespace QgsWms
...
+       * \since QGIS 3.34
```
Comment (pblottiere) [inline src/server/services/wms/qgsmaprendererjobproxy.h:57]:
```suggestion
```

File: src/server/qgsfcgiserverresponse.cpp [hunk 10]
```diff
@@ -23,15 +23,137 @@
...
+    QgsMessageLog::logMessage( QStringLiteral( "FCGI_stdin is null! Socket monitoring disable." ),
+                               QStringLiteral( "FCGIServer" ),
+                               Qgis::MessageLevel::Warning );
+  }
+#endif
+}
+
+void SocketMonitoringThread::run( )
+{
+  if ( ! mIpcFd )
```
Comment (pblottiere) [inline src/server/qgsfcgiserverresponse.cpp:96]:
```suggestion
  if ( mIpcFd < 0 )
```

File: src/server/qgsfcgiserverresponse.cpp [hunk 11]
```diff
@@ -23,15 +23,137 @@
...
+                               QStringLiteral( "FCGIServer" ),
+                               Qgis::MessageLevel::Warning );
+    return;
+  }
+
+#ifdef Q_OS_LINUX
+  char c;
+  while ( !*mIsResponseFinished )
+  {
+    ssize_t x = recv( mIpcFd, &c, 1, MSG_PEEK | MSG_DONTWAIT ); // see https://stackoverflow.com/a/12402596
```
Comment (pblottiere) [inline src/server/qgsfcgiserverresponse.cpp:108]:
```suggestion
    const ssize_t x = recv( mIpcFd, &c, 1, MSG_PEEK | MSG_DONTWAIT ); // see https://stackoverflow.com/a/12402596
```

File: src/server/qgsfcgiserverresponse.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (benoitdm-oslandia) [reply src/server/qgsfcgiserverresponse.cpp:26]: indeed

File: src/server/qgsfcgiserverresponse.cpp [hunk 12]
```diff
@@ -23,15 +23,137 @@
 #include <fcgi_stdio.h>
 #include <QDebug>
 
+#include <unistd.h>
+#include "qgslogger.h"
+
+#ifdef Q_OS_LINUX
```
[the remark this answers is not in the document]
Comment (benoitdm-oslandia) [reply src/server/qgsfcgiserverresponse.cpp:29]: fixed

File: src/server/qgsfcgiserverresponse.cpp (hunk 1 above)
Comment (benoitdm-oslandia) [reply src/server/qgsfcgiserverresponse.cpp:36]: fixed

File: src/server/qgsfcgiserverresponse.cpp (hunk 3 above)
Comment (benoitdm-oslandia) [reply src/server/qgsfcgiserverresponse.cpp:104]: fixed

File: src/server/qgsfcgiserverresponse.cpp (hunk 4 above)
Comment (benoitdm-oslandia) [reply src/server/qgsfcgiserverresponse.cpp:70]: fixed

File: src/server/qgsfcgiserverresponse.h (hunk 6 above)
Comment (benoitdm-oslandia) [reply src/server/qgsfcgiserverresponse.h:37]: fixed

File: src/server/qgsfcgiserverresponse.h (hunk 7 above)
Comment (benoitdm-oslandia) [reply src/server/qgsfcgiserverresponse.h:54]: fixed

File: src/server/services/wms/qgsmaprendererjobproxy.h (hunk 8 above)
Comment (benoitdm-oslandia) [reply src/server/services/wms/qgsmaprendererjobproxy.h:56]: fixed

**Comment by pblottiere** [comment]
Hello @benoitdm-oslandia,

Can you rebase please? Then I'll merge this PR.

**Comment by benoitdm-oslandia** [comment]
@pblottiere done!

**Comment by benoitdm-oslandia** [comment]
@pblottiere PR was rebased, minGW build is still failed for non linked reason. Does it look good to you?

**Review by pblottiere — APPROVED** [review]

**Comment by pblottiere** [comment]
> Does it look good to you?

Yes :). Thanks @benoitdm-oslandia :+1:

**Comment by benoitdm-oslandia** [comment]
Thanks @pblottiere!

**Pull request merged**